src/main.rs

```diff
@@ -1609,7 +1613,12 @@ async fn main() {
 
     loop {
         tokio::select! {
-            credentials = async { discovery.as_mut().unwrap().next().await }, if discovery.is_some() => {
+            credentials = async {
+                match discovery.as_mut() {
+                    Some(d) => d.next().await,
+                    _ => None
```

Comment:
While a `None` value seems harmless, really all of these should output an error level message and possibly exit with an error. If `None` is returned here, that means that the sending side of the channel has been dropped. This is unrecoverable and will leave `librespot` in a state of limbo. Things _may_ continue to work (particulary in the case of discovery and player events) but at least be unexpected and nontransparent to the user.

Reply:
No. If discovery is disabled from the get go discovery.as_mut() will be None. It's perfectly OK for discovery to be none so long as there are cached creds.

Reply:
My mistake I was thinking about the unwrapping the call to `recv`.

Reply:
I don't think `spirc_task` should ever be `None` though.

Reply:
If you want a spot to cut and run it should be at:
```rust
warn!("Discovery stopped!");
```

That should be made into a (better) error and we should exit after that.

Reply:
And:
```rust
                        error!("Discovery stopped unexpectedly");
                        exit(1);
```

Reply:
And for the sake of less panics I also put else branches on the `tokio::select!`'s. They don't really do much but the docs say:

> The select! macro panics if all branches are disabled and there is no provided else branch.

Reply:
> My mistake I was thinking about the unwrapping the call to recv.


Well it's either unwrap or:
```rust
                match player_event_channel.as_mut() {
                    Some(p) => p.recv().await,
                    _ => None
                }
```

Or something like this:
```rust

                match player_event_channel.as_mut() {
                    Some(p) => p.recv().await,
                    _ => {
                        error!("Player event channel closed unexpectedly");
                        exit(1);
                    }
                }
```

Reply:
> I don't think spirc_task should ever be None though.

It may never be I don't know but it's either unwrap or something like I've got:

```rust
                    if let Some(task) = spirc_task.as_mut() {
                        task.await;
                    }
```

Reply:
Scratch that. If it will never be None then why is there a bunch of code dedicated to recovering from a shutdown?

Reply:
Let's leave that in. Right now I can't think of when that code would be triggered. As-is `spirc` can panic but not exit gracefully, I think. However, that's all about the change in `new-api` and this may then come in handy.